Support additional device types

nairbv opened the pull request:
Summary: Adding support for additional device types besides cuda and cpu.

Differential Revision: D10175683

ezyang commented on aten/src/ATen/core/Device.cpp:
```diff
@@ -18,7 +18,32 @@ std::pair<Device::Type, size_t> parse_type(const std::string& device_string) {
+  if (position != std::string::npos) {
+    return {Device::Type::MKLDNN, 6};
+  }
+  position = device_string.find("mkldnn");
```
I think you didn't want to test this twice.

nairbv replied:
fixed

ezyang reviewed (changes requested):
Duplicate test

goldsborough commented on aten/src/ATen/core/Device.cpp:
```diff
@@ -18,7 +18,28 @@ std::pair<Device::Type, size_t> parse_type(const std::string& device_string) {
     return {Device::Type::CUDA, 4};
   }
-  AT_ERROR("Expected 'cpu' or 'cuda' device type at start of device string");
+  position = device_string.find("mkldnn");
```
I'm not a fan of this. This was fine when it was two strings, but now this whole function could be reduced by probably 75% if we use a map. Having to count the the length of the string we're looking for is also super fragile. Again, fine when two strings, but now this function warrants a rewrite

apaszke replied:
I think this code still has an opportunity to be faster than an `std` map

goldsborough replied:
It could be done with `std::array`, then the overhead should be barely noticeable. I just think everything that changes from `if` clause to `if` clause should be aggregated in some container, and then the logic can be separated from the config, in a sense

nairbv replied:
I'm still slowly remembering C++, but agree this isn't ideal. I'm submitting an update using boost, let me know what you think.

goldsborough replied:
Oh we don't have boost in OSS unfortunately

nairbv replied:
ah... ok. already submitted but will discuss offline.

goldsborough reviewed (changes requested):
Sorry but I think this could be better

ezyang commented on aten/src/ATen/core/Device.cpp:
```diff
@@ -46,29 +57,21 @@ std::pair<Device::Type, size_t> parse_type(const std::string& device_string) {
 // }
 Device::Device(const std::string& device_string) : Device(Type::CPU) {
-  AT_CHECK(!device_string.empty(), "Device string must not be empty");
+  std::istringstream ss(device_string);
```
Errr.... I had no problems with the old code, but if we're going to optimize, isn't stringstream/getline one of those C++ standard library functions that is famously slow?

apaszke replied:
Yeah, that will be super slow. It will copy `device_string`, and is really not needed in our simple use case.

ezyang commented:
For the record: I am OK with landing either the old code, or the new code, once its build failures are fixed.

nairbv committed:
Support additional device types (#12293)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/12293

Adding support for additional device types besides cuda and cpu.

Reviewed By: ezyang

Differential Revision: D10175683

fbshipit-source-id: 0b9c7ee9285b762fb268fed439ab2c9f7d37c3c5
